### DISCUSSION TITLE
Does IFC contain Data in STEP format?

### DISCUSSION THREAD

**Discussion by Mannshoch** [IfcOpenShell/IfcOpenShell#2710] [General]
I read the Wiki page for IFC
https://en.wikipedia.org/wiki/Industry_Foundation_Classes#File_Formats
It seems that there are references to STEP.

Is it correct that the core of IFC is a STEP file?
If that so may converting to .stp could be speedup much more.

**Comment by aothms** [comment]
IFC uses several standards from the step (iso 10303) family of standards such as the schema definition language (express) and the exchange formats (.ifc file - 10303 p21) and some of the geometry definitions (IfcExtrudedAreaSolid vs extruded_area_solid). A conversion is not trivial because the schemas (the available classes) are completely different.

**Reply by Mannshoch** [reply]
Do you mean IFC classes or something model dependent?  If I convert data to STP I'm normally not interested in **any** other data than the bodies itself. If a body has an identical geometry like some other, but the Information behind that body is different, remove it and thread it as copy into the step.

Currently a 5MB ifc get blown up to a 300MB STEP.

**Reply by aothms** [reply]
The IfcConvert .STP exporter has little to no intelligence or attempt at trying to output a minimal file. Data will blow up because of (a) no product-level geometry reuse (b) fully explicit data (extrusion -> brep) (c) maybe also 2d (edge parametrizations) and 3d data.

**Reply by Mannshoch** [reply]
We normally request specific bodies, for the work. But if we get an IFC mostly the whole Building is included. That's the cause why I have to  insist to get another filetype like sat or step
I developed an assumption. Could it be, that If an architect export an IFC from selected bodies, their tool export the building as tesselation and the selected body as geometric body? 

If that's the cause I need an "export only geometric body" settings.

**Reply by Moult** [reply]
If you're only interested in geometry, have you considered other formats? GLB? DAE? OBJ?

The BlenderBIM Add-on can load most IFCs very quickly, then if you're only interested in geometry you can use Blender to cull whatever you need pretty easily.

**Reply by Mannshoch** [reply]
From the filetypes ifcconvert support only STP and SAT is usable for me. The CAD I use do not support GLB, DAE and only in bad Quality OBJ.
I have the Idea to create a directory on our NAS that is whatched by Inotify and auto-convert Ifc I put into it to STP or SAT.

**Reply by Mannshoch** [reply]
I'm still interested in a new cli setting to exclude tessellated geometry for converting. Should I create an Issue for that?

**Reply by Moult** [reply]
I don't think it makes sense in the C++ side to add such a specific setting. Maybe a more generic setting would be considered, but why don't you quickly just patch it in Python prior to conversion?

**Reply by Mannshoch** [reply]
Because I'm not a developer. I'm only user. I have some basic experience in cmd and bash. That's it.

Could you pleas elaborate, why you think this is not a generic use? 

> We normally request specific bodies, for the work. But if we get an IFC mostly the whole Building is included. That's the cause why I have to insist to get another filetype like sat or step I developed an assumption. Could it be, that If an architect export an IFC from selected bodies, their tool export the building as tesselation and the selected body as geometric body?
> 
> If that's the cause I need an "export only geometric body" settings.

From my point of view with the customer we have. It seems common that exported IFC has tessellation beside the geometry which is only for Information. That made it very normal for me to use it every time with such a setting. If it's available.

**Reply by Moult** [reply]
I think there is some confusion in terminology. You're trying to convert only geometry that aren't tessellations (e.g. faceted breps, face based surface models, and tesselated face sets)? Instead, you only want to convert stuff like CSG solids and extruded solids?

Not only does this mean half your bodies will be missing (in IFC, the "body" can be either a mesh-like or a solid-like geometry), it also is no different from the perspective of IfcConvert because as @aothms mentioned the STEP converter deals with fully explicit data.

**Reply by Mannshoch** [reply]
> Instead, you only want to convert stuff like CSG solids and extruded solids?

Yes,

> Not only does this mean half your bodies will be missing..

That is what I want. I assume that the IFC I receive contains CSG solids and extruded solids, precisely because these data were intended for transmission. Whether other surface types play a for a geometric data transfer, I do not know exactly. I'm not able to post an example so we can not figure out exactly what kind of geometry is for display only and not for data transfer. The data I receive is not owned by my.

> as @aothms mentioned the STEP converter deals with fully explicit data.

I' do not understand what that means. Sorry.

**Reply by Moult** [reply]
It means that any intelligence or parametric data like extrusions will be reduced back to tessellations in STP. Doesn't that defeat the purpose of what you're after?

**Reply by Mannshoch** [reply]
Yes and No.
In my case to use IfcConvert STEP files as base for new geometry. Connection Points like pipes and Walls are important. As long the tolerance is not to bad, it's usable. 
I wish this feature to reduce the file size of the step because my CAD could not handle a 12 floor Building with all details if I simply need one room. That's to much data for importing. And If the cad is able to do so, I sometimes wait for more than one Coffee to drink.

Main purpose is to reduce the content to the geometry I need. If at the end these Files are Tessellation data itself is only in the point of tolerance important. Until now, I have no problem with it.

By the way:
A workaround could be to split the IFC into a smaller IFC with only the geometric data. Because generally my CAD is able to handle IFC but has problem with to much data. So my CAD could convert the files at the end.

**Reply by Moult** [reply]
You could try the IfcPatch recipes to extract only what you need, I think that's a more practical approach than separating by geometry paradigm. https://blenderbim.org/docs-python/ifcpatch.html https://blenderbim.org/docs-python/autoapi/ifcpatch/recipes/index.html https://blenderbim.org/docs-python/autoapi/ifcpatch/recipes/SplitByBuildingStorey/index.html https://blenderbim.org/docs-python/autoapi/ifcpatch/recipes/ExtractElements/index.html
